metricflow/api/metricflow_client.py

```diff
@@ -0,0 +1,246 @@
+from typing import Dict, List, Optional, Tuple
+
+from metricflow.configuration.config_handler import ConfigHandler
+from metricflow.configuration.constants import CONFIG_DWH_SCHEMA
+from metricflow.dataflow.sql_table import SqlTable
+from metricflow.engine.metricflow_engine import (
+    MetricFlowEngine,
+    MetricFlowExplainResult,
+    MetricFlowQueryRequest,
+    MetricFlowQueryResult,
+)
+from metricflow.engine.models import Dimension, Materialization, Metric
+from metricflow.engine.utils import build_user_configured_model_from_config, convert_to_datetime
+from metricflow.model.model_validator import ModelValidator
+from metricflow.model.semantic_model import SemanticModel
+from metricflow.model.validations.validator_helpers import ValidationIssueType
+from metricflow.protocols.sql_client import SqlClient
+from metricflow.sql.optimizer.optimization_levels import SqlQueryOptimizationLevel
+from metricflow.sql_clients.sql_utils import make_sql_client_from_config
+
+
+class MetricFlowClient:
+    """MetricFlow Python Interface."""
+
+    def __init__(
+        self,
+        sql_client: Optional[SqlClient] = None,
+        semantic_model: Optional[SemanticModel] = None,
+        system_schema: Optional[str] = None,
```

Comment:
This still bothers me a bit - I don't like exposing more than we need to since changing public APIs is quite difficult to do safely.

In thinking about this a bit more, it certainly makes sense that we'd want to expose the SqlClient and system schema as overrides, but I'm still not sold on the SemanticModel.  What's the use case for this?

I also think we might benefit from splitting this into an initializer with these elements required and a factory method that takes no arguments and builds those parameters from a config. Then developers get their ease of use and we can document the initializer as something that we don't want people to invoke directly. It also keeps the config fetching separate from the initialization of the object itself, which I think is desirable.

Reply:
These overrides are here as just an alternative where one could put the values directly into the constructor itself versus putting the values into the config file. For example, a user may want to make use of this client, but for whatever reason they don't want to use the config file or something. Then they would pass the required params to the initializer, which would directly allow them to use it without having to create/fill out the config file.

With that being said, I don't know how important this use case is, but we could technically simplify this further and instead of passing a SemanticModel object we can have them pass the `path_to_models_directory` as a str and we build the SemanticModel for them.

Reply:
So I can definitely see scenarios for passing a custom SqlClient in - like if someone implements that interface for an engine we don't support inside MetricFlow itself - it's just having this open in the initializer with a magic fallback to the config that worries me.

I'd prefer to either:

1.  Make these parameters required and have the config mechanism handled by a factory method, then document that factory method  as the standard way to initialize one of these things
2. Remove these parameters and only support the config approach, and then provide a separate MetricFlowClient class that requires these overrides

I think option 1 is cleanest and simplest, because the config initialization is a separate convenience rather than something we'd ordinarily bake into the client directly.

Reply:
so, I drafted up 2 ways of going at this,
1. using a staticmethod in the class to instantiate
```python
@staticmethod
def from_config() -> MetricFlowClient:
    """Builds a MetricFlowClient via config yaml file."""
    handler = ConfigHandler()
    sql_client = make_sql_client_from_config(handler)
    user_configured_model = build_user_configured_model_from_config(handler)
    schema = handler.get_value(CONFIG_DWH_SCHEMA)

    return MetricFlowClient(
        sql_client=sql_client,
        user_configured_model=user_configured_model,
        system_schema=schema,
    )

def __init__(
    self,
    sql_client: SqlClient,
    user_configured_model: UserConfiguredModel,
    system_schema: str,
):
   pass
```
2. using a factory class
```python
class ConfigInitMetricFlowClient(MetricFlowClient):
    def __init__(self) -> None:
        handler = ConfigHandler()
        sql_client = make_sql_client_from_config(handler)
        user_configured_model = build_user_configured_model_from_config(handler)
        schema = handler.get_value(CONFIG_DWH_SCHEMA)
        super().__init__(
            sql_client=sql_client, user_configured_model=user_configured_model, system_schema=system_schema
        )


class MetricFlowClient:
    """MetricFlow Python Interface."""

    def __init__(
        self,
        sql_client: SqlClient,
        user_configured_model: UserConfiguredModel,
        system_schema: str,
    ):
    pass
```

Any preference here?

Reply:
implemented 1)

Reply:
For posterity, we discussed separately and decided the static method was cleaner.